docs/extensions/payment-processors/create.md

```diff
@@ -1,5 +1,227 @@
+
+## Introducing `PropertyBag` objects
+
+Currently as of 5.19 `$params` is a sprawling array with who-knows-what keys. From 5.21, a better way is to pass in a `Civi\Payment\PropertyBag` object instead.
+
+!!! danger "todo"
+    update 5.21 to actual value
+
+This object has getters and setters that enforce standardised property names and a certain level of validation and type casting. For example, the property `contactID` (note capitalisation) has `getContactID()` and `setContactID()`.
+
+For backwards compatibility, this class implements `ArrayAccess` which means if old code does `$propertyBag['contact_id'] = '123'` or `$propertyBag['contactID'] = 123` it will translate this to the new `contactID` property and use that setter which will ensure that accesing the property returns the integer value *123*. When this happens deprecation messages are emitted to the log file. New code should not use array access.
+
+### Checking for existence of a property
+
+Calling a getter for a property that has not been set will throw a `BadMethodCall` exception.
+
+Code can require certain properties by calling
+`$propertyBag->require(['contactID', 'contributionID'])` which will throw an `InvalidArgumentException` if any property is missing. These calls should go at the top of your methods so that it's clear to a developer.
+
+You can check whether a property has been set using `$propertyBag->has('contactID')` which will return `TRUE` or `FALSE`.
+
+### Multiple values, e.g. changing amounts
+
+All the getters and setters take an optional extra parameter called `$label`. This can be used to store two (or more) different versions of a property, e.g. 'old' and 'new'
+
+```php
+<?php
+use Civi\Payment\PropertyBag;
+//...
+$propertyBag = new PropertyBag();
+$propertyBag->setAmount(1.23, 'old');
+$propertyBag->setAmount(2.46, 'new');
+//...
+$propertyBag->getAmount('old'); // 1.23
+$propertyBag->getAmount('new'); // 2.46
+$propertyBag->getAmount(); // throws BadMethodCall
+```
+
+This means the value is still validated and type-cast as an amount (in this example).
+
+### Custom payment processor-specific data
+
+!!! danger "todo"
+    This is not yet implemented in core.
+
+Sometimes a payment processor will requrie custom data. e.g. Stripe may use a `paymentMethodID` and a `paymentIntentID` on its `doPayment()` method.
+
+Property names should be prefixed, e.g. `stripe_paymentMethodID` and set using `PropertyBag->setCustomProperty($prop, $value, $label = 'default')`.
+
+The payment class is responsible for validating such data; anything is allowed by `setCustomProperty`, including `NULL`.
+
+Where support for custom properties is needed for a method, e.g. `doPayment()`, you should implement a method as follows:
+
+```php
+<?php
+use Civi\Payment\PropertyBag;
+use Civi\Payment\Exception\PaymentProcessorException;
+
+class CRM_Core_Payment_MyExtension extends CRM_Core_Payment {
```

Comment:
I don't mind extending CRM_Core_Payment, it's the overriding of its methods or properties that I would be concerned about, insofar as they don't invoke parent's or change the results of invoking parent's.

Reply:
My bad.